src/connectors/onchain_events/mod.rs

```diff
@@ -0,0 +1,361 @@
+                add_event(
+                    fid,
+                    EventType::IdRegister(IdRegisterEvent::ChangeRecovery {
+                        recovery_address: recovery,
+                    }),
+                );
+                return Ok(());
+            }
+            Some(&KeyRegistryAbi::Add::SIGNATURE_HASH) => {
+                let KeyRegistryAbi::Add {
+                    fid,
+                    key: _,
+                    keytype,
+                    keyBytes,
+                    metadatatype,
+                    metadata,
+                } = event.log_decode()?.inner.data;
+                let fid = Uint::to::<u64>(&fid);
+                add_event(
+                    fid,
+                    EventType::Signer(SignerEvent::Add {
+                        key: keyBytes,
+                        key_type: keytype,
+                        metadata,
+                        metadata_type: metadatatype,
+                    }),
+                );
+                return Ok(());
+            }
+            Some(&KeyRegistryAbi::Remove::SIGNATURE_HASH) => {
+                let KeyRegistryAbi::Remove {
+                    fid,
+                    key: _,
+                    keyBytes,
+                } = event.log_decode()?.inner.data;
+                let fid = Uint::to::<u64>(&fid);
+                add_event(
+                    fid,
+                    EventType::Signer(SignerEvent::Remove { key: keyBytes }),
+                );
+                return Ok(());
+            }
+            Some(&KeyRegistryAbi::AdminReset::SIGNATURE_HASH) => {
+                let KeyRegistryAbi::AdminReset {
+                    fid,
+                    key: _,
+                    keyBytes,
+                } = event.log_decode()?.inner.data;
+                let fid = Uint::to::<u64>(&fid);
+                add_event(
+                    fid,
+                    EventType::Signer(SignerEvent::AdminReset { key: keyBytes }),
+                );
+                return Ok(());
+            }
+            Some(&KeyRegistryAbi::Migrated::SIGNATURE_HASH) => {
+                let KeyRegistryAbi::Migrated { keysMigratedAt } = event.log_decode()?.inner.data;
+                let migrated_at = Uint::to::<u64>(&keysMigratedAt);
+                add_event(0, EventType::SignerMigrated { migrated_at });
+                return Ok(());
+            }
+            _ => return Ok(()),
+        }
+    }
+
+    pub async fn run(&mut self) -> Result<(), SubscribeError> {
+        // Subscribe to new events starting from now.
+        let filter = Filter::new().address(vec![STORAGE_REGISTRY, KEY_REGISTRY, ID_REGISTRY]);
+        let subscription = self.provider.watch_logs(&filter).await?;
+        let mut stream = subscription.into_stream();
+        while let Some(events) = stream.next().await {
+            for event in events {
+                let result = self.process_log(&event).await;
+                if result.is_err() {
```

Comment:
nit: I (guess!) the idiomatic way would be to use `match` here. But I've be writing rust for only a few days! And I don't know if we have any established code conventions.